### DISCUSSION TITLE
Is there any easy way to access vector of activations present in value / policy heads for given chess position

### DISCUSSION THREAD

**Discussion by int8** [LeelaChessZero/lc0#1932] [Q&A]
I have PGN/fen (or any format of chess position) and I want to pass it forward through value/policy network of lc0 to get activation vector(s) - any hint on how to achieve that ?

**Comment by Naphthalin** [comment]
There are ways people have done this -- best to join the Lc0 Discord server and ask there. The GitHub issues are only for actual issues with the Lc0 engine binary.

**Comment by int8** [comment]
thanks - on my way :)

**Comment by int8** [comment]
Ok I will actually leave my comment here in case anyone is looking for solution in the future: 

most likely steps in the right direction: 

- convert chosen model to ONNX via `lc0 leela2onnx` command 
- use `onnx` python package to modify converted model to include intermediate layer as output 

```
import onnx
model = onnx.load("converted_model.onnx")
print([n.name for n in model.graph.node]) # this will list all intermediate nodes 
# now let's say I want to read values from `/block9/conv2/relu` (layer that goes to policy/value heads):

from onnx import ValueInfoProto
intermediate_layer_value_info = ValueInfoProto()
intermediate_layer_value_info.name = '/block9/conv2/relu'
model.graph.output.append(intermediate_layer_value_info)
model.graph.output.extend([intermediate_layer_value_info])
model.save("modified_model.onnx") 
```

- load modified model 

```
import onnxruntime as ort
ort_sess = ort.InferenceSession('modified_model.onnx')
outputs = ort_sess.get_outputs()
print([o.name for o in outputs]) # now output will include intermediate layer 
```

- run inference (this is tricky because it requires preparation of  input for the model, at this point I will need mapping of fen into that  input w.r.t  https://github.com/LeelaChessZero/lc0/blob/master/src/neural/encoder.cc#L134 (which is in progress) 



I will try to update this issue

**Comment by bastiansg** [comment]
Hi! @int8!

I'm very interested in this topic; I wanted to know if you managed to encode a specific chess position using the Leela Chess Zero neural network.

**Comment by mooskagh** [comment]
One of the ways to encode the position is to use python bindings as described here: https://github.com/LeelaChessZero/lc0/pull/1261.

It's also not that hard to encode it manually. The mentioned function https://github.com/LeelaChessZero/lc0/blob/master/src/neural/encoder.cc#L134 has some bits of complicated logic, but they all were used for experimental input plane encodings. All modern networks use INPUT_CLASSICAL_112_PLANE encoding which is fairly simple (and I believed it's exactly the same as in the original AlphaZero paper).

In any case, it will be faster to find solution in our Discord chat at http://lc0.org/chat.
